### ISSUE TITLE
Assign a unique ID to each Component

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1192]
### What is a unique ID?
A [universally unique identifier (UUID)](https://en.wikipedia.org/wiki/Universally_unique_identifier) is a 128-bit value like `123e4567-e89b-12d3-a456-426655440000`. UUIDs can be generated randomly with near certainty of uniqueness using
- [Graeme Hill's CrossGuid library](https://github.com/graeme-hill/crossguid) (uses MIT license)
- [Boost's "Uuid" library](http://www.boost.org/doc/libs/1_61_0/libs/uuid/uuid.html) (but requires Boost)
- ["UuidCreate" on Windows](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379205(v=vs.85).aspx) (but not cross-platform)

### What problems would be solved?
The current API relies on user-modifiable names (string properties) to form connections. This design follows most naturally from the 3.3 API, but introduces some potential concerns:
 1. Components on the same path must have unique names at all times. This requirement puts constraints on the design of the GUI and the user experience (e.g., automatically suffix top-level Component name with "_1" after copying or delay uniqueness check) and exposes the underlying API architecture/data structures to the GUI-level user.
 2. Renaming a Component invalidates connections.
 3. Components must be named. Perhaps the name of the Component is unimportant to the user (e.g., if creating and connecting mathematical operation blocks using the future GUI model builder).
 4. Component names cannot contain special characters (see #596). Currently, we (should) disallow ".", "/", "\", " "; a path-handling class could help (see #1069), but we would still need to consider serialization/deserialization of special characters.

### What is the proposal?
 1. Assign a UUID to each Component upon creation (and deserialization, if none exists).
 2. The UUID would be constant so connections would be unaffected by Component renaming. (Modifying properties like "name" and "comment" should not trigger `finalizeFromProperties()`, which should be reserved for modifications that would affect the behavior of the Model.)
 3. The UUID would be serialized, which could help users extend models in xml (e.g., by inserting the UUID to form connections rather than assembling full paths).
 4. It might be helpful for the GUI (now and for eventual graphical model building) to hold onto a unique ID tag for each Component, regardless of its name, path, or connection status.
 5. Continue supporting lookup by name and full path for backwards compatibility and usability, but map to UUID behind the scenes.

### See also
[Simulink objects are assigned "Simulink Identifiers" (SIDs)](http://www.mathworks.com/help/simulink/ug/overview_bso67kd.html)

**Comment by aymanhab** [comment]
@tkuchida Very nicely summarized and articulated :+1:. 

I'd add to the Problems Solved:
1. The mechanism is not specific to "Components" so any object can have a uuid (e.g. Appearance/Material or other OpenSim objects that don't fit the Component abstraction fully).
2. In contexts where multiple instances of a Model could be used (e.g. GUI) ComponentPathName may never be unique and we jump through hoops to make that possible, this could be elimintaed using uuid.
3. For Visualizer interface, right now I have to use PathName to map Components to Visuals, the latter map to uuid since it is used by the Visualizer, working around previous item, all this will be simplified and code eliminated.

**Comment by aseth1** [comment]
A UUID would be helpful to uniquely identify Components and thereby permit reliable serialization of connections. We should move that way for these reasons.
We should also keep in mind that name (path name) based connections cannot be made to disappear overnight:
1. XML editing is the primary method of building models. Manipulating names, even though it is somewhat error prone due to typos, etc..., the context is easy to understand and easy to debug if you have the wrong name.  If the API issues UUIDs (at construction or deserialization) then how do you create a new Model in XML? Using names.
2. We do not have graphical tools for model building and to identify relationships. A UUID would be a reliable way of serializing the reference, for sure, but I think we are far from having graphical model building.
3. We have to support existing based dependencies because all our models are name based.

There is an underlying simplicity to the current API that I would like to preserve. Properties define the model.  Edits to a property is a change to the model and it should invalidate the model (its system) until all property changes are finalized. A name change would not change the UUID and so should reconnect correctly (if connections are based on UUID). This is one of the more compelling reasons to use the UUID - so we can treat names as properties and not as an identifier.

The main challenge is how to manage a name based scheme for XML editing and existing models AND a UUID-based connection scheme for the benefits listed going forward. For example, I have an existing model file, say leg6dof9musc with a device I added via XML.  Upon loading the .osim file each Component now has a UUID.  To "wire" up the model I need a `connectee_id` (instead of `connectee_name`) so now we must:
1. still use `connectee_name` to locate a dependency (like we have now) and satisfy the `Connector`

2. store the UUID on the `Connector` for serialization

**Comment by aymanhab** [comment]
I agree that the current GUI doesn't allow for graphical model building or workflow building. That said let me explain how three.js uses UUIDs to manage its scene-graph. A Scene has a flat list of materials and another for geometry each with a unique uuid then a tree that represents the scene-graph where each node has its own uuid and refers to geometry and material from the previous two lists using uuid, Ownership/scope issues do not exist.  The analogy to OpenSim would be a flat list (library) of components with unique uuids then a tree that refers to these components to form a model hierarchically.  I think we should design for the ultimate goal and minimize the time we spend on perfecting intermediate solutions.

**Comment by aseth1** [comment]
> let me explain how three.js uses UUIDs to manage its scene-graph. A Scene has a flat list of materials and another for geometry each with a unique uuid then a tree that represents the scene-graph where each node has its own uuid and refers to geometry and material from the previous two lists using uuid, Ownership/scope issues do not exist

I believe the proposal at hand is to replace the use and our dependency on pathname for UUIDs which I completely agree is a good idea. 

Your implied proposal to use the three.js approach for managing all Components in OpenSim, while interesting and likely to have its advantages, represents a complete re-architecture of OpenSim and current concept of Components. Surely, there are also pitfalls for switching to an uber flat list. I disagree that their are no ownership/scope issues, they are just placed elsewhere. If you delete a Component today, all of its constituent components are also deleted. In a flat scheme where no component owns any other they either remain unused or you have to implement specialized code that removes all its subcomponents from the list. I am not saying it is not possible or even difficult, it just needs to be addressed. It also a major change to our property system which yields ownership by default. Even if we decided this was the way to  go long term,  these changes are not going to happen for 4.0.

 > I think we should design for the ultimate goal and minimize the time we spend on perfecting intermediate solutions.

I agree, however intermediate solutions (I think you are referring to the use of pathnames as the identifier) still have to work and be robust. There is no other way to identify dependencies in existing model's XML. In the API you can find Components using filtered lists or search by name and type and then connect components directly. These are capabilities the GUI could use for model building as well. The path name is primarily for serialization and I agree switching to UUID would be more robust here too.  We could switch to UUIDs today if we didn't have the requirement for users to edit XML to build models.  Adding the UUID is not the primary challenge - it is supporting user edits of our serialized model files. 

We could allow users to edit  `conectee_IDs` directly for newly serialized models, but since we would still have to support models without UUIDs to begin with, we still need to use pathnames. It will also generally be easier to debug names in XML than IDs so I can see users preferring to use pathnames in XML (I could be wrong). 

The worst case scenario, in my mind, is to have both pathnames and IDs appear in XML that users are still expected to edit. We then need rules like: IDs trump pathnames and pathnames are used if no ID is present; if users want to edit and use pathnames they have to manually wipeout the associated ID, and so on. Alternatively, we switch to IDs completely (no pathnames) load old models, convert paths to IDs.  Then we require users to use IDs and in the case of adding new Components they must assign a unique ID themselves. 

We must have plans to deal with these issues if we push to use UUIDs in 4.0.
